library/BerlinPipeline/src/Juvix/BerlinPipeline/Env.hs

```diff
@@ -37,23 +39,65 @@ newtype EnvS b = EnvS (State T b)
 
 data StopADT = Stop
 
+-- | Register the pipeline function to the environment
 registerStep :: CircularList.T Step.Named -> EnvS ()
-registerStep = notImplemented
+registerStep l = do
+  modify@"registeredPipeline" $ \i -> i <> l
 
+-- | Create a named group of pipeline steps or nested grouping of pipeline steps.
 defPipelineGroup :: NameSymbol.T -> [CircularList.T Step.Named] -> CircularList.T Step.Named
-defPipelineGroup = notImplemented
+defPipelineGroup sym ls = foldl' (<>) (CircularList.init sym) ls
 
+-- | Tell the environment to stop at a particular step when running the environment.
 stopAt :: NameSymbol.T -> EnvS ()
-stopAt = notImplemented
+stopAt sym = put@"stoppingStep" (Just sym)
 
 stopAtNothing :: EnvS ()
-stopAtNothing = notImplemented
-
-eval :: T -> Pipeline.ComputationalInput
-eval = notImplemented
+stopAtNothing = put@"stoppingStep" Nothing
 
-run :: EnvS b -> T -> Pipeline.ComputationalInput
-run = notImplemented
+-- Change our environment
+eval :: MonadIO m => T -> m Pipeline.CIn
+eval
+  ( T
+      input@(Pipeline.CIn wEnv@(Pipeline.WorkingEnv sexp context) surr)
+      pipeline@(CircularList.T reclist)
+      stoppingStep
+    ) = do
+    case nextStep of
+      Nothing -> pure input
+      Just (CircularList.NonCircSchema nStep) -> do
+        let name = Step.name nStep
+        let namedInput = Pipeline.nameCIn name input
+        if shouldStop stoppingStep name
+          then pure input
+          else liftIO $ do
```

Comment:
We probably don't want the `liftIO` here, as it seems overally wide scoping over the actual logic